include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1143,27 +1144,113 @@ struct __early_exit_find_or
 // parallel_find_or - sync pattern
 //------------------------------------------------------------------------
 
+template <typename Tag>
+struct __parallel_find_or_n_groups_tuner
+{
+    // Calculate the number of work groups.
+    //
+    // @param _ExecutionPolicy&& __exec - policy
+    // @param std::size_t __n_groups - the source amount of work-groups
+    // @param std::size_t __wgroup_size - the source work-group size
+    // @param std::size_t __rng_n - the source data size
+    // @return std::size_t - evaluated number of work-groups
+    template <typename _ExecutionPolicy>
+    std::size_t
+    operator()(_ExecutionPolicy&& /*__exec*/, std::size_t __n_groups, const std::size_t /*__wgroup_size*/,
+               const std::size_t /*__rng_n*/) const
+    {
+        return __n_groups;
+    }
+};
+
+template <>
+struct __parallel_find_or_n_groups_tuner<oneapi::dpl::__internal::__device_backend_tag>
+{
+    // Calculate the number of work groups.
+    //
+    // @param _ExecutionPolicy&& __exec - policy
+    // @param std::size_t __n_groups - the source amount of work-groups
+    // @param std::size_t __wgroup_size - the source work-group size
+    // @param std::size_t __rng_n - the source data size
+    // @return std::size_t - evaluated number of work-groups
+    template <typename _ExecutionPolicy>
+    std::size_t
+    operator()(_ExecutionPolicy&& /*__exec*/, std::size_t __n_groups, const std::size_t __wgroup_size,
+               const std::size_t __rng_n) const
+    {
+        // No tuning for FPGA_EMU
+#if _ONEDPL_FPGA_EMU
+        return __n_groups;
+#endif
+
+        // If all source data fits into one work-group, then we need only one work-group
+        if (__rng_n <= __wgroup_size)
+            return 1;
+
+        // Size: [ 16'777'216, ..,  67'108'864 ) -> minimum number of iterations per work-item is 164
+        // Size: [  4'194'304, ..,  16'777'216 ) -> minimum number of iterations per work-item is  32
+        // Size: [  1'048'576, ...,  4'194'304 ) -> minimum number of iterations per work-item is  16
+        // Size: [    262'144, ...,  1'048'576 ) -> minimum number of iterations per work-item is   8
+        if (__n_groups > 1)
+        {
+            constexpr std::array<std::size_t, 6> __lower_bounds_of_sizes         = { 262'144, 1'048'576, 4'194'304, 16'777'216, 67'108'864 };
```

Comment:
Do we understand why this gives us a performance boost?  
Perhaps it is some combination of a minimum level of occupancy mixed with sufficiently large number of iterations to get maximal benefit from early exit opportunities...

I'm asking because I'm wondering if there is a more targeted way to reach this same result, as this includes many empirically tuned numbers. Reducing the number of these magic values, or even better, relying upon some query-able device characteristics would be more maintainable certainly.

Reply:
"I'm asking because I'm wondering if there is a more targeted way to reach this same result, as this includes many empirically tuned numbers. Reducing the number of these magic values, or even better, relying upon some query-able device characteristics would be more maintainable certainly."
 - unfortunately I haven't any ideas about this at current time.

Reply:
We can simple drop this PR, it's give us not more then 10% profit.

Reply:
I'm not necessarily suggesting to drop it, I just wonder if there is a more direct way of going about it. 

The method before the PR was to just use the number of compute units as the number of groups.  This makes sense as it allows us to fill the device, but also have large wgs to maximize early exit.

Our benchmarks have a good variety of data configurations, including cases with no data to find at all.  However, I'm wondering this is tuning too much to the benchmark data. I think our basic understanding is that there is some penalty from underutilizing the hardware (by progressive factors of 2), which we are weighing against the expected savings we have from potential to early exit by increasing the iters per workitem by progressive factors of 2.

I would expect that we would want to target either: 
(a) a specific number of work groups
-or-
(b) a specific number of iters per work item

Beyond that I fear we might be somehow training to our benchmarking data, unless we have a compelling reasoning behind it.

Reply:
I agree with this. I think the speedup on PVC shows that there is a better launch configuration. I think we should first understand why this improves the runtime and then come up with some calculation potential based of the number of compute units. Also, we should benchmark this on different hardware architectures to not introduce slowdowns.

I made a quick prototype to use iters per work item with powers of two, but this was worse. I suspect that the tuned n values don't match anymore because we end up with a different number of groups than in the current approach. So, collecting data for the iterations or tuning the number of groups might be better suited here.

Reply:
Please see my comment at https://github.com/oneapi-src/oneDPL/pull/1723#issuecomment-2250116248